### PR TITLE
feat: migrate smart contract v4

### PR DESCRIPTION
**Pull request by QYuQianchen**
### Description

This PR supercedes https://github.com/hoprnet/hoprnet/pull/5984

Key features:

- Introduce single step node-safe on-boarding
- Update Channel events
- Remove unused isHoprNodeManagementModule check in module inclusion
- Use Ownable2Step to transfer ownership in key contracts.
- Add ownership in NodeStakeFactory
- Use UUPS upgradable proxy patterns in module cration
- Add keyid in Keybindings
- Add migration contract to facilitate users to migrate from v3 to v4
- Allow ERC777 token recipient hook by default
- Include announcement contract as target in module initiation
- Remove ticketIndexOffset in channels to save storage
- Make HoprLedger properties public
- Correct HoprLedger snapshot interval to seconds
- Cleanup unnecessary approve hash in Node Safe creation
- Move NetworkRegistry to static and remove them from contract deployment workflow

Miscellaneous:
- Bump forge-std to v.10.0, OpenZeppelin to 5.4.0
- Include files for code audit in Aug 2024
- Updated Solidity compiler to version 0.8.30

CI:
- Temporarily disable gas report generation, in favor of sufficient resource for unit tests

### Deployment
v4 Rotsee network contracts:
```
{
        "announcements": "0xe45A8DBDADafd86AB7e2368b4749864083331cFC",
        "channels": "0x9680F21d4583ad27F2e6bc3005EDF120FFf057B7",
        "module_implementation": "0x79C3bF06E96A9373765284aFb0a35e9529E2B3F2",
        "node_safe_migration": "0xe9670B5D87c87111C9050D915971B73b9f5021a9",
        "node_safe_registry": "0xaFa257f2799835D2E29e7eC7ee448530F9d8Cb20",
        "node_stake_factory": "0xcB0841cc3DBDE97aC52B945F02Ee4F3D8707d977",
        "ticket_price_oracle": "0xFA251d4C367683d6181531afd5964E660aCf43A0",
        "token": "0xD4fdec44DB9D44B8f2b6d529620f9C0C7066A2c1",
        "winning_probability_oracle": "0xa641822a52AcbDc0c0123337f715C1d9756c21bD"
}
```

## Summary by CodeRabbit

* **New Features**
  * On-chain key-binding support and per-node multiaddr lookup in announcements
  * Safe/module migration workflow and deterministic safe/module deployment
  * Batch node inclusion/funding (multi-node operations)
  * Channel events now emit compact channel state for improved observability

* **Bug Fixes**
  * Event signatures, ticket/index validation and domain-separator corrections
  * Ownership/initialization and transfer-edge case fixes

* **Documentation**
  * New detailed audit reports and migration guides

* **Tests**
  * Expanded test coverage and new mocks for migration, key bindings, channels

### CODE REVIEW

File: ethereum/contracts/test/utils/Crypto.sol [hunk 1]
```diff
@@ -17,7 +18,6 @@ abstract contract CryptoUtils is Test, HoprCrypto, SECP2561k {
         address dest;
         uint256 amount;
         uint256 maxTicketIndex;
-        uint256 indexOffset;
         uint256 epoch;
```
[the remark this answers is not in the document]
Comment (QYuQianchen) [reply ethereum/contracts/test/utils/Crypto.sol]: Thanks. Missing documentation will be tackled altogether. Please create a sub-issue for this under https://github.com/hoprnet/contracts/issues/6

File: ethereum/contracts/test/node-stake/permissioned-modules/NodeManagementModule.t.sol [hunk 2]
```diff
@@ -850,7 +954,7 @@ contract HoprNodeManagementModuleTest is
             channels, Clearance.FUNCTION, TargetType.CHANNELS, TargetPermission.BLOCK_ALL, defaultFunctionPermission
         ); // clerance: FUNCTION default ALLOW_ALL
         stdstore.target(address(moduleProxy)).sig("owner()").checked_write(safe);
-        vm.mockCall(channels, abi.encodeWithSignature("token()"), abi.encode(token));
+        vm.mockCall(channels, abi.encodeWithSignature("TOKEN()"), abi.encode(token));
```
[the remark this answers is not in the document]
Comment (QYuQianchen) [reply ethereum/contracts/test/node-stake/permissioned-modules/NodeManagementModule.t.sol]: Thanks. Missing documentation will be tackled altogether. Please create a sub-issue for this under https://github.com/hoprnet/contracts/issues/6

File: ethereum/contracts/src/Crypto.sol [hunk 3]
```diff
@@ -676,6 +682,9 @@ abstract contract HoprCrypto {
                 let b0Payload := mload(0x40)
 
                 // payload[0..KECCAK256_BLOCKSIZE] = 0
+                for { let i := 0 } lt(i, KECCAK256_BLOCKSIZE) { i := add(i, 0x20) } {
+                    mstore(add(b0Payload, i), 0)
+                }
```
[the remark this answers is not in the document]
Comment (QYuQianchen) [reply ethereum/contracts/src/Crypto.sol]: Memory should be set to zero and not calldata

File: ethereum/bindings/src/codegen/hopr_node_safe_registry_events.rs [hunk 4]
```diff
@@ -0,0 +1,789 @@
...
+    /// The runtime bytecode of the contract, as deployed on the network.
+    ///
+    /// ```text
+    ///0x
+    /// ```
+    #[rustfmt::skip]
+    #[allow(clippy::all)]
+    pub static DEPLOYED_BYTECODE: alloy_sol_types::private::Bytes = alloy_sol_types::private::Bytes::from_static(
+        b"",
+    );
```
[the remark this answers is not in the document]
Comment (QYuQianchen) [reply ethereum/bindings/src/codegen/hopr_node_safe_registry_events.rs]: Bindings are autogen. Please skip code reviews for files under `codegen` folder

File: ethereum/bindings/src/codegen/hopr_ticket_price_oracle_events.rs [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+    /// The runtime bytecode of the contract, as deployed on the network.
+    ///
+    /// ```text
+    ///0x
+    /// ```
+    #[rustfmt::skip]
+    #[allow(clippy::all)]
+    pub static DEPLOYED_BYTECODE: alloy_sol_types::private::Bytes = alloy_sol_types::private::Bytes::from_static(
+        b"",
+    );
```
[the remark this answers is not in the document]
Comment (QYuQianchen) [reply ethereum/bindings/src/codegen/hopr_ticket_price_oracle_events.rs]: Bindings are autogen. Please skip code reviews for files under `codegen` folder

File: ethereum/bindings/src/codegen/hopr_winning_probability_oracle_events.rs [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+            }
+        }
+        impl WinProb {
+            /// The Solidity type name.
+            pub const NAME: &'static str = stringify!(@ name);
+            /// Convert from the underlying value type.
+            #[inline]
+            pub const fn from_underlying(
+                value: alloy::sol_types::private::primitives::aliases::U56,
+            ) -> Self {
```
[the remark this answers is not in the document]
Comment (QYuQianchen) [reply ethereum/bindings/src/codegen/hopr_winning_probability_oracle_events.rs]: Bindings are autogen. Please skip code reviews for files under `codegen` folder

File: ethereum/bindings/src/codegen/hopr_winning_probability_oracle_events.rs [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+            #[inline]
+            fn stv_abi_packed_encoded_size(&self) -> usize {
+                <alloy::sol_types::sol_data::Uint<
+                    56,
+                > as alloy_sol_types::SolType>::abi_encoded_size(self)
+            }
+        }
+        impl WinProb {
+            /// The Solidity type name.
+            pub const NAME: &'static str = stringify!(@ name);
```
[the remark this answers is not in the document]
Comment (QYuQianchen) [reply ethereum/bindings/src/codegen/hopr_winning_probability_oracle_events.rs]: Bindings are autogen. Please skip code reviews for files under `codegen` folder

**Comment by Teebor-Choka** [comment]
@copilot @coderabbitai review this PR from an audit perspective as well. The code changed here is outside of the `vendor` directory, no need to audit that. Also, suggest improvements and fixes that could be performed on the existing implementation and propose where documentation could be improved, because it is unclear.

File: ethereum/bindings/src/bin/hopr-contract-addresses.rs [hunk 8]
```diff
@@ -8,48 +8,32 @@
...
-//!         "network_registry": "0x15a315E1320cFF0de84671c0139042EE320CE38d",
-//!         "network_registry_proxy": "0x20559cbD3C2eDcD0b396431226C00D2Cd102eB3F",
-//!         "node_safe_registry": "0x4F7C7dE3BA2B29ED8B2448dF2213cA43f94E45c0",
-//!         "node_stake_v2_factory": "0x791d190b2c95397F4BcE7bD8032FD67dCEA7a5F2",
-//!         "token": "0xD4fdec44DB9D44B8f2b6d529620f9C0C7066A2c1",
-//!         "ticket_price_oracle": "0x442df1d946303fB088C9377eefdaeA84146DA0A6",
-//!         "winning_probability_oracle": "0xC15675d4CCa538D91a91a8D3EcFBB8499C3B0471"
-//!       }
-//!     },
+//!      "addresses": {
```
[the remark this answers is not in the document]
Comment (Teebor-Choka) [reply ethereum/bindings/src/bin/hopr-contract-addresses.rs]: The existing rotsee and dufour are not v4 smart contracts compatible, and therefore would be moved to readme.md into the deprecated section.

File: ethereum/bindings/src/codegen/hopr_ticket_price_oracle_events.rs [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+    #[derive(serde::Serialize, serde::Deserialize)]
+    #[derive(Default, Debug, PartialEq, Eq, Hash)]
+    /**Event with signature `TicketPriceUpdated(uint256,uint256)` and selector `0xd4c5e06b1ae097ba02372652a7adaa6e4a8e00be527497a3ad0ebc3f761ef3fb`.
```
[the remark this answers is not in the document]
Comment (QYuQianchen) [reply ethereum/bindings/src/codegen/hopr_ticket_price_oracle_events.rs]: Bindings are autogen. Please skip code reviews for files under codegen folder
